Mathlib/Data/QPF/Multivariate/Basic.lean

```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2018 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Avigad, Simon Hudon
+
+! This file was ported from Lean 3 source module data.qpf.multivariate.basic
+! leanprover-community/mathlib commit dc6c365e751e34d100e80fe6e314c3c3e0fd2988
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.PFunctor.Multivariate.Basic
+
+/-!
+# Multivariate quotients of polynomial functors.
+
+Basic definition of multivariate QPF. QPFs form a compositional framework
+for defining inductive and coinductive types, their quotients and nesting.
+
+The idea is based on building ever larger functors. For instance, we can define
+a list using a shape functor:
+
+```lean
+inductive ListShape (a b : Type)
+| nil : ListShape
+| cons : a -> b -> ListShape
+```
+
+This shape can itself be decomposed as a sum of product which are themselves
+QPFs. It follows that the shape is a QPF and we can take its fixed point
+and create the list itself:
+
+```lean
+def List (a : Type) := fix ListShape a -- not the actual notation
+```
+
+We can continue and define the quotient on permutation of lists and create
+the multiset type:
+
+```lean
+def Multiset (a : Type) := QPF.quot List.perm List a -- not the actual notion
+```
+
+And `Multiset` is also a QPF. We can then create a novel data type (for Lean):
+
+```lean
+inductive Tree (a : Type)
+| node : a -> Multiset Tree -> Tree
+```
+
+An unordered tree. This is currently not supported by Lean because it nests
+an inductive type inside of a quotient. We can go further and define
+unordered, possibly infinite trees:
+
+```lean
+coinductive Tree' (a : Type)
+| node : a -> multiset Tree' -> Tree'
+```
+
+by using the `cofix` construct. Those options can all be mixed and
+matched because they preserve the properties of QPF. The latter example,
+`Tree'`, combines fixed point, co-fixed point and quotients.
+
+## Related modules
+
+ * constructions
+   * fix
+   * cofix
+   * quot
+   * comp
+   * sigma / pi
+   * prj
+   * const
+
+each proves that some operations on functors preserves the QPF structure
+
+## Reference
+
+! * [Jeremy Avigad, Mario M. Carneiro and Simon Hudon, *Data Types as Quotients of Polynomial Functors*][avigad-carneiro-hudon2019]
+-/
+
+
+universe u
+
+open MvFunctor
+
+/-- Multivariate quotients of polynomial functors.
+-/
+class MvQPF {n : ℕ} (F : TypeVec.{u} n → Type _) [MvFunctor F] where
+  p : MvPFunctor.{u} n
```

Comment:
```suggestion
  P : MvPFunctor.{u} n
```
I would write this one as a capital P, since that is how we commonly refer to polynomial functors